src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,407 @@
+            continue
+
+        try:
+            test_file_analysis = analyse_test_file(file_path)
+            code_file_analysis = analyse_code_file(file_path)
+
+            # Combine the analysis results with the original data
+            analysis_result = {
+                "file_path": file_path,
+                **test_file_analysis,  # This unpacks the dictionary test_file_
+                # analysis and includes all its key-value pairs in the analysis_
+                # result dictionary.
+                **code_file_analysis,
+            }
+
+            batch_results.append(analysis_result)
+
+        except ValueError as e:
+            logger.error(f"Failed to analyse {file_path}: {str(e)}")
+            # Add failed file to results with error indicators
+            analysis_result = {
+                "file_path": file_path,
+                "num_test_cases": -1,
+                "num_assertions": -1,
+                "has_setup": False,
+                "has_teardown": False,
+                "complexity": -1,
+                "cyclomatic_complexity": -1,
+                "lines_of_code": -1,
+                "num_functions": -1,
+                "analysis_error": str(e),
+            }
+            batch_results.append(analysis_result)
+            continue  # Continue with next file
+
+        except Exception as e:
+            logger.error(f"Unexpected error analysing {file_path}: {str(e)}")
+            # Add failed file to results with error indicators
+            analysis_result = {
+                "file_path": file_path,
+                "num_test_cases": -1,
+                "num_assertions": -1,
+                "has_setup": False,
+                "has_teardown": False,
+                "complexity": -1,
+                "cyclomatic_complexity": -1,
+                "lines_of_code": -1,
+                "num_functions": -1,
+                "analysis_error": f"Unexpected error: {str(e)}",
+            }
+            batch_results.append(analysis_result)
+            continue  # Continue with next file
+
+        # Save the batch results every BATCH_SIZE files
+        if len(batch_results) >= BATCH_SIZE:
+            batch_df = pd.DataFrame(batch_results)
+            final_df = pd.concat([final_df, batch_df], ignore_index=True)
+            final_df.to_csv(output_file, index=False)
+            logger.info(f"Saved analysis results for {len(batch_results)} files")
+
+            # Clear the batch results after saving
+            batch_results.clear()
+
+    # Save any remaining results after the loop
+    if batch_results:
+        batch_df = pd.DataFrame(batch_results)
+        final_df = pd.concat([final_df, batch_df], ignore_index=True)
+
+        # Remove duplicates before saving
+        final_df.to_csv(output_file, index=False)
+        logger.info(
+            f"Saved final batch of analysis results for {len(batch_results)}" f"files"
```

Comment:
Minor: add a space before the word `files`.

Reply:
This is done.